samples/media/ultrahdr/src/main/java/com/example/platform/media/ultrahdr/video/UltraHDRToHDRVideo.kt

```diff
@@ -0,0 +1,629 @@
+ * limitations under the License.
+ */
+
+package com.example.platform.media.ultrahdr.video
+
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.graphics.ColorSpace
+import android.graphics.HardwareBufferRenderer
+import android.graphics.HardwareBufferRenderer.RenderResult
+import android.graphics.RenderNode
+import android.hardware.DataSpace
+import android.hardware.HardwareBuffer
+import android.media.Image
+import android.media.ImageWriter
+import android.media.MediaCodec
+import android.media.MediaCodecInfo
+import android.media.MediaCodecList
+import android.media.MediaFormat
+import android.media.MediaMuxer
+import android.net.Uri
+import android.os.Build
+import android.os.Bundle
+import android.os.Handler
+import android.os.HandlerThread
+import android.util.Log
+import android.view.LayoutInflater
+import android.view.Surface
+import android.view.View
+import android.view.ViewGroup
+import android.widget.ImageView
+import android.widget.Toast
+import androidx.annotation.RequiresApi
+import androidx.fragment.app.Fragment
+import androidx.media3.common.MediaItem
+import androidx.media3.common.Player
+import androidx.media3.exoplayer.ExoPlayer
+import com.bumptech.glide.Glide
+import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
+import com.example.platform.media.ultrahdr.databinding.UltrahdrToHdrVideoBinding
+import com.google.android.catalog.framework.annotations.Sample
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.time.Duration
+import java.util.UUID
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+
+
+@Sample(
+    name = "UltraHDR to HDR Video",
+    description = "This sample demonstrates converting a series of UltraHDR images into a HDR " +
+            "video." + "The sample leverages GPU hardware acceleration to render and encode the " +
+            "images.",
+    documentation = "https://developer.android.com/guide/topics/media/hdr-image-format",
+    tags = ["UltraHDR"],
+)
+@RequiresApi(Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
+class UltraHDRToHDRVideo : Fragment() {
+    /**
+     *  Android ViewBinding.
+     */
+    private var _binding: UltrahdrToHdrVideoBinding? = null
+    private val binding get() = _binding!!
+
+    /**
+     * [ImageWriter], used to produce [Image] objects that will to convert UltraHDR image data to an
+     * HDR color space (in this sample, HLG) to then be encoded by the [MediaCodec]
+     */
+    private lateinit var imageWriter: ImageWriter
+
+    /**
+     * The ImageWriters [HandlerThread] & [Handler], used to process
+     * [ImageWriter.setOnImageReleasedListener] callbacks.
+     */
+    private lateinit var imageWriterHandler: Handler
+    private val imageWriterThread: HandlerThread by lazy {
```

Comment:
Is it possible to use coroutines instead? This sample mixes coroutines with threads and it's a bit confusing.

Reply:
Unfortunately, `ImageWriter` and `MediaCodec.Callback` need their own handlers to process callback calls on thier own thread.

Not sure if or when they will switch to coroutine scopes if ever.

Reply:
Not familiar with this APIs but I think if you pass null to the handler it will use the thread handler
https://developer.android.com/reference/android/media/ImageWriter#setOnImageReleasedListener(android.media.ImageWriter.OnImageReleasedListener,%20android.os.Handler)

If you then call this method within a coroutine using a Dispatcher.IO shouldn't that work?

In anycase shouldn't you stop the thread on stop/destroy?

Reply:
You can create a `Dispatcher` from a `Handler` using `Handler.asCoroutineDispatcher()`